slides lecture 1 first draft

Connortsui20 opened the pull request:
First draft of lecture 1 slides

Compile with VSCode's Marp extension

jurplel commented on week1/slides.md:
```diff
@@ -0,0 +1,852 @@
+
+# Why Rust?
+
+TODO A few slides about this...
```
There is great stuff from the old slide deck on this.
high-level:
<img width="568" alt="image" src="https://github.com/rust-stuco/lectures/assets/10716694/191937d2-1022-4846-acd6-f072a3187f06">

Definitely should mention concurrency too, I think that is the most convincing reason to switch

Connortsui20 replied:
Yeah I've allocated 10 minutes to this so we'll go in depth

jurplel commented on week1/slides.md:
```diff
@@ -0,0 +1,852 @@
+
+---
+
+# Rust seems awesome!
```
I don't really like this slide. It seems kinda pointless, or at the very least boring.

Connortsui20 replied:
im removing it

jurplel commented on week1/slides.md:
```diff
@@ -0,0 +1,852 @@
+    * `src/main.rs`
+    * `Cargo.toml`
+    * `.git` repository and `.gitignore`
+* We will come back to those last two in the future
```
i don't see the need to discuss gitignore

Connortsui20 replied:
i changed this in the next commit

jurplel commented on week1/slides.md:
```diff
@@ -0,0 +1,852 @@
+* This creates an executable file at `target/debug/hello_cargo`
+* To build for release, run `cargo build --release`
+* What if we want to actually run this executable?
+    * Can run `./target/debug/hello_cargo --`, but this is verbose
```
"this is verbose" made me think that the output would somehow be verbose---maybe just say "this is a lot to type" or something

jurplel commented on week1/slides.md:
```diff
@@ -0,0 +1,852 @@
+# Shadowing vs Mutability
+
+* We get a compile time error if we try to modify a non-`mut` variable
+* Using `let` multiple times allows a few transformations on a value but keep it immutable
```
Please explain this line? I think it is sufficient to say "We can use let multiple times to shadow the old names". It isn't accurate to say we have any transformations occurring here, we are just making a new variable with the same name in a more deeply nested scope.

Connortsui20 replied:
I'm thinking we say:
* Mutability lets us change the value of a variable
    * We get a compile time error if we try to modify a non-`mut` variable
* Shadowing allows us to change what a variable logically represents
    * In addition to changing the value, so it can also change types

thoughts? I will put it in the next set of commits

jurplel commented on week1/slides.md:
````diff
@@ -0,0 +1,852 @@
+let mut spaces = "   ";
+spaces = spaces.len();
+```
+* The second one does not compile!
````
I feat that if we trick them like this, we won't do a good job of explaining the difference between shadowing and mutability. I can be persuaded otherwise

jurplel commented on week1/slides.md:
```diff
@@ -0,0 +1,852 @@
+
+# Data Types
+
+Like most langauges, there are two main types of Types.
```
categories of types would be less redundant here

jurplel commented on week1/slides.md:
````diff
@@ -0,0 +1,852 @@
+```
+
+* Unlike tuples, all elements must be the same type.
+* The number of elements is always fixed (stack allocated)
````
Add here: "These arrays correspond to arrays you know and (maybe) love from C"

jurplel commented on week1/slides.md:
````diff
@@ -0,0 +1,852 @@
+```
+
+* Rust will check if the index is in bounds at runtime
+   * This is _not_ done in C/C++
````
Add as a bullet point: There is a way to index into an array without this runtime cost, if you really want to

Connortsui20 replied:
that is something we'll say in lecture, but we don't need to put in slides.
technically this is not always true since the compiler will optimize this away a lot of the time but again we can say that in lecture rather than making the slides verbose

jurplel commented on week1/slides.md:
```diff
@@ -0,0 +1,852 @@
+
+# Functions
+
+Like all languages, Rust has functions.
```
*most languages

Connortsui20 replied:
what languages don't have functions???

jurplel replied:
You can make it all with an asterisk if you want. If we put all we're going to get heckled "what about brainfuck" or smth i swear

jurplel commented on week1/slides.md:
```diff
@@ -0,0 +1,852 @@
+fn main() {
+    let y = {
+        let x = 3;
+        x + 1
```
i think this is very confusing. I suggest you show how to return from a function before talking about statements/expressions. It seems like statements/expressions are unmotivated until we get here, and then the students have to keep thinking about functions while being taught about statements and try to put everything together. This just feels really messy, I think.

Connortsui20 replied:
i agree

I went in the exact order of the rust book to make it easier on the first draft, we can move functions in front of this

jurplel replied:
👍perfect

jurplel commented on week1/slides.md:
````diff
@@ -0,0 +1,852 @@
+```
+
+* Functions must have a specific return value or none at all
+    * No return type is equivalent to the unit type `()`
````
add something like: (There is no void, more aligned with actual type theory!)

Connortsui20 replied:
Again, this is something we can say in lecture but we don't need on the slides

jurplel commented on week1/slides.md:
```diff
@@ -0,0 +1,852 @@
+
+* Functions must have a specific return value or none at all
+    * No return type is equivalent to the unit type `()`
+* Notice again that there is no semicolon after `x + 1`
```
Yeah, as I was saying I think the ordering of the slides makes this kinda awkward. I can really see the students being confused about returning values at this point. I would *strongly* prefer we do return before statements

jurplel commented on week1/slides.md:
```diff
@@ -0,0 +1,852 @@
+}
+
+fn plus_one(x: i32) -> i32 {
+    x + 1
```
You should introduce using the `return` keyword here as well!

jurplel commented on week1/slides.md:
```diff
@@ -0,0 +1,852 @@
+
+![bg right:25% 75%](../images/ferris_does_not_compile.svg)
+
+`if` expressions must take a boolean.
```
to be more precise, must *condition* on a boolean expression.

also, add: unlike C/C++/Python

jurplel commented on week1/slides.md:
````diff
@@ -0,0 +1,852 @@
+}
+```
+
+* `break` and `continue` apply to the innnermost loop where they are called
````
maybe add:
equivalent to `while (true)`

or say it out loud anyway

Connortsui20 replied:
yes will say it out loud

jurplel commented on week1/slides.md:
```diff
@@ -0,0 +1,852 @@
+
+---
+
+Do we talk about loop lables
```
I suggest we put it in the appendix, i.e. at the end of the slide deck

Connortsui20 replied:
i think that we should still show it as something you can do. I added stuff in the next commit which is unfortunately just a code example, but we don't have to talk about it much. The point is that we just make them aware that it is something they can do if they ever run into needing it. Then, they can look it up themselves.

jurplel replied:
sounds good to me

jurplel commented on week1/slides.md:
```diff
@@ -0,0 +1,852 @@
+        println!("{number}!");
+    }
+    println!("LIFTOFF!!!");
+}
```
Maybe a bullet point: "No C-style `for` loops!"

Connortsui20 replied:
will say out loud

jurplel commented on week1/slides.md:
```diff
@@ -0,0 +1,852 @@
+
+---
+
+# Course logistics
```
This could also go at the end, we should discuss

Connortsui20 replied:
We should definitely do it at the end

jurplel replied:
We should make sure not to run out of time, then

jurplel commented on week1/slides.md:
```diff
@@ -0,0 +1,852 @@
+
+---
+
+# Installing Rust
```
A slide that should definitely be at the end:

Here is the next homework! You will have one of these per week

1CoolDavid commented on week1/slides.md:
If we're going to discuss constants should we discuss the rules of what can and cannot be a constant value?

Connortsui20 replied:
since we're only bringing up the basic data types here I don't think we need to.

jurplel reviewed (commented):
Really nice work! I am quite happy with how the plan came together and this seems like a good technology for slide creation, rather than powerpoint or beamer.

I left some comments, i only have one serious concern, otherwise mostly small stuff

Connortsui20 commented on week1/slides.md:
Which part? I assume you're talking about the scopes

Connortsui20 commented on week1/slides.md:
```diff
@@ -0,0 +1,852 @@
+paginate: true
+---
+
+# Intro to Rust
```
What did you have in mind? We can easily add a fancy intro slide by getting the image and pasting it into the markdown

jurplel replied:
I can spin something up. I want to make some nice graphic design, i will throw together a website before this.

Connortsui20 commented on week1/slides.md:
```diff
@@ -0,0 +1,852 @@
+
+![bg right:35% 65%](../images/ferris.svg)
+
+---
```
I think that we should definitely do most of this last

a quick introduction of ourselves and a brief overview of course goals can go at the beginning, i was thinking in this order:
- intro us 3
- why rust
- course goals

jurplel replied:
sure, sounds good

jurplel commented:
Let me know when you are ready for another review.

Connortsui20 commented:
The main thing that needs to be changed here is that functions need to come before scopes.

Connortsui20 commented:
Changed the ordering of the function section to introduce function return without `return` before expressions and scopes, and then a revisit of functions for return types.

Also added some divider slides and cleaned up some other stuff.

Still need to do the beginning and the end

jurplel commented:
- Make sure you cover concurrency as a positive
- For course goals:
	- Instead of proficiency equal to any other language (kinda unrealistic?), maybe “Become a proficient Rust programmer, to an intermediate or expert level”
	- eliminate usage of the word such for the importance bullet point

- Cargo is not built into Rust, it Is included with Rust distributions
- Might want to mention cmake along with pip and nom, since cmake is probably exactly what rust devs were trying to outdo
- “To check if your code compiles, use cargo check” -> “To check your code for errors, you can use cargo check” (we say check for compile on a later bullet point)
- On the shadowing slides, “what a variable logically represents” -> “what data a variable’s name (symbol) refers to”
- Typo on the bullet point below that “so it can also change types”
- Maybe remove the bullet point “what are types though?” I think we should say that out loud
- I think you should write “We can also omit the return keyword here” instead for the return keyword omission part.
- What answer are you expecting from “Why are we allowed to do this?”
- You did not define statements/expressions in a precise way. We should discuss. https://doc.rust-lang.org/beta/reference/statements-and-expressions.html I don’t have time to get into it right now.
- Maybe change the phrasing “none at all” for function return types, it’s kinda awkward imo
- loop will keep looping -> Loop will loop forever (Maybe mention its like while true in other langs)
- I like the liftoff example
- The for loop section seems a little slim. Maybe want to mention that we’ll cover collections more in depth later. Maybe mention that ranges are collections or something. Maybe collections isn’t the most precise language either… You should mention you can’t do C-style for loops. You should show an example of a range loop without .rev(), it’ll be too confusing for first presenting it—show it without parens too.
-last page typo, programming assignments.
- Mention feedback on communication slide
- I think we should say less than an hour instead of 30-45 minutes. I think we should say “if you are spending more than that” instead of more than an hour.
- Well done!

Connortsui20 commented on week1/slides.md:
```diff
@@ -31,8 +30,8 @@ Talk a bit about ourselves
-* Ferris is Rust's mascot
-* Ferris will be our friend throughout this semester
+* Ferris is Rust's mascot, and ours too!
+* Ferris will appropriately express joy and anguish throughout this semester
```
I don't think we need to put this in

jurplel replied:
Sure, remove it

Connortsui20 commented on week1/slides.md:
```diff
@@ -65,9 +64,9 @@ https://www.youtube.com/watch?v=DnT-LUQgc7s
 # What is Rust?
 
-Developed originally by Mozilla for "systems programming," now community-driven and open-source.
+Developed originally by Mozilla to replace C++ code in Firefox, now community-driven and open-source.
```
Is this actually true? Even if it is I don't think it was the original goal

jurplel replied:
Not quite
> Rust grew out of a personal project begun in 2006 by [Mozilla](https://en.wikipedia.org/wiki/Mozilla) Research employee Graydon Hoare.[[19]](https://en.wikipedia.org/wiki/Rust_(programming_language)#cite_note-MITTechReview-22) Mozilla began sponsoring the project in 2009 as a part of the ongoing development of an experimental browser engine called [Servo](https://en.wikipedia.org/wiki/Servo_(software)),[[20]](https://en.wikipedia.org/wiki/Rust_(programming_language)#cite_note-infoq2012-23) which was officially announced by Mozilla in 2010.[[](https://en.wikipedia.org/wiki/Rust_(programming_language)#cite_note-MattAsay-24)

It's a pretty good simplification I think, to justify its purpose. "Systems programming" isn't very specific, and this is why Mozilla sponsored it basically (considering they built Servo in it and then imported Servo components into Gecko)

Connortsui20 commented on week1/slides.md:
I don't think the affine type system is a modern feature, it is very specifically a rust feature though

jurplel replied:
I agree

Connortsui20 commented on week1/slides.md:
```diff
@@ -185,8 +187,8 @@ Requires thinking a lot up front before you write down code
 * Rust is only 8 years old
-* Much smaller ecosystem compared to older langauges
-* _Gaining adoption quickly_
+* Smaller/less mature ecosystem compared to some older langauges
```
Smaller and less mature

jurplel replied:
I do not understand the comment. Are you suggesting switching from smaller/less mature -> smaller and less mature? If so, good w/ me

Connortsui20 commented on week1/slides.md:
```diff
@@ -195,9 +197,8 @@ Requires thinking a lot up front before you write down code
 
 * Compile time is slow
-* Vendor support for large C++ libraries
-* Windows support
```
Windows support is still a pitfall

jurplel replied:
What is wrong with Windows support?

Connortsui20 replied:
The same issue with any open source project - most people use linux and mac so windows bindings are relatively young. This may change however now that microsoft is investing heavily into rust

Connortsui20 replied:
we'll leave it out anyways

Connortsui20 reviewed (commented):
a few changes

Connortsui20 commented:
small thing, but next year we should considering using implicit named arguments in `println`, as it seems the rust community is moving towards those by default. But for now, i'm going to make them all explicit for consistency.
